File: BackEnd/Java/springshop_maven/src/main/java/com/pbshop/springshop/pcbuilder/PcBuilderService.java
package com.pbshop.springshop.pcbuilder;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.pbshop.springshop.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.springshop.common.exception.BusinessException;
import com.pbshop.springshop.common.exception.ErrorCode;
import com.pbshop.springshop.pcbuilder.dto.PcBuilderDtos.AddPartRequest;
import com.pbshop.springshop.pcbuilder.dto.PcBuilderDtos.CompatibilityRuleRequest;
import com.pbshop.springshop.pcbuilder.dto.PcBuilderDtos.CreateBuildRequest;
import com.pbshop.springshop.pcbuilder.dto.PcBuilderDtos.UpdateBuildRequest;
import com.pbshop.springshop.product.Product;
import com.pbshop.springshop.product.ProductRepository;
import com.pbshop.springshop.user.User;
import com.pbshop.springshop.user.UserRepository;

@Service
@Transactional
public class PcBuilderService {
    private final PcBuildRepository pcBuildRepository;
    private final PcBuildPartRepository pcBuildPartRepository;
    private final PcCompatibilityRuleRepository ruleRepository;
    private final ProductRepository productRepository;
    private final UserRepository userRepository;
    private final ObjectMapper objectMapper;

    public PcBuilderService(PcBuildRepository pcBuildRepository, PcBuildPartRepository pcBuildPartRepository,
            PcCompatibilityRuleRepository ruleRepository, ProductRepository productRepository, UserRepository userRepository,
            ObjectMapper objectMapper) {
        this.pcBuildRepository = pcBuildRepository;
        this.pcBuildPartRepository = pcBuildPartRepository;
        this.ruleRepository = ruleRepository;
        this.productRepository = productRepository;
        this.userRepository = userRepository;
        this.objectMapper = objectMapper;
    }

    @Transactional(readOnly = true)
    public Map<String, Object> listMine(AuthenticatedUserPrincipal principal) {
        User user = requireCurrentUser(principal);
        List<Map<String, Object>> items = pcBuildRepository.findByUserIdOrderByIdDesc(user.getId()).stream().map(this::toSummary).toList();
        return page(items);
    }
    public Map<String, Object> create(AuthenticatedUserPrincipal principal, CreateBuildRequest request) {
        User user = requireCurrentUser(principal);
        PcBuild build = new PcBuild();
        build.setUser(user);
        build.setName(request.name());
        build.setDescription(request.description());
        return toDetail(pcBuildRepository.save(build));
    }
    @Transactional(readOnly = true)
    public Map<String, Object> show(Long id) {
        PcBuild build = requireBuild(id);
        build.setViewCount(build.getViewCount() + 1);
        return toDetail(build);
    }
    public Map<String, Object> update(AuthenticatedUserPrincipal principal, Long id, UpdateBuildRequest request) {
        PcBuild build = requireOwnedBuild(principal, id);
        if (request.name() != null && !request.name().isBlank()) { build.setName(request.name()); }
        if (request.description() != null || request.description() == null) { build.setDescription(request.description()); }
        return toDetail(build);
    }
    public Map<String, Object> delete(AuthenticatedUserPrincipal principal, Long id) {
        pcBuildRepository.delete(requireOwnedBuild(principal, id));
        return Map.of("message", "견적이 삭제되었습니다.");
    }
    public Map<String, Object> addPart(AuthenticatedUserPrincipal principal, Long buildId, AddPartRequest request) {
        PcBuild build = requireOwnedBuild(principal, buildId);
        Product product = productRepository.findById(request.productId())
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "상품을 찾을 수 없습니다."));
        PcBuildPart part = new PcBuildPart();
        part.setPcBuild(build);
        part.setPartType(request.partType());
        part.setProduct(product);
        part.setQuantity(request.quantity() == null ? 1 : request.quantity());
        pcBuildPartRepository.save(part);
        return toDetail(build);
    }
    public Map<String, Object> removePart(AuthenticatedUserPrincipal principal, Long buildId, Long partId) {
        PcBuild build = requireOwnedBuild(principal, buildId);
        PcBuildPart part = pcBuildPartRepository.findByIdAndPcBuildId(partId, build.getId())
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "부품을 찾을 수 없습니다."));
        pcBuildPartRepository.delete(part);
        return toDetail(build);
    }
    @Transactional(readOnly = true)
    public Map<String, Object> compatibility(Long buildId) {
        PcBuild build = requireBuild(buildId);
        List<PcBuildPart> parts = pcBuildPartRepository.findByPcBuildIdOrderByIdAsc(build.getId());
        List<Map<String, Object>> issues = ruleRepository.findAllByOrderByIdAsc().stream()
                .filter(rule -> parts.stream().anyMatch(part -> part.getPartType().equalsIgnoreCase(rule.getSourcePartType()))
                        && parts.stream().anyMatch(part -> part.getPartType().equalsIgnoreCase(rule.getTargetPartType())))
                .filter(rule -> "CATEGORY_MISMATCH".equalsIgnoreCase(rule.getRuleType()))
                .map(rule -> {
                    Map<String, Object> item = new LinkedHashMap<>();
                    item.put("ruleId", rule.getId());
                    item.put("message", rule.getName());
                    item.put("severity", "WARNING");
                    return item;
                })
                .toList();
        return Map.of("compatible", issues.isEmpty(), "issues", issues);
    }
    public Map<String, Object> share(AuthenticatedUserPrincipal principal, Long buildId) {
        PcBuild build = requireOwnedBuild(principal, buildId);
        if (build.getShareCode() == null || build.getShareCode().isBlank()) {
            build.setShareCode(UUID.randomUUID().toString().replace("-", "").substring(0, 12));
        }
        return Map.of("shareUrl", "/api/v1/pc-builds/shared/" + build.getShareCode(), "shareCode", build.getShareCode());
    }
    @Transactional(readOnly = true)
    public Map<String, Object> shared(String shareCode) { return toDetail(pcBuildRepository.findByShareCode(shareCode)
            .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "공유 견적을 찾을 수 없습니다."))); }
    @Transactional(readOnly = true)
    public Map<String, Object> popular() { return page(pcBuildRepository.findAllByOrderByViewCountDescIdDesc().stream().map(this::toSummary).toList()); }
    @Transactional(readOnly = true)
    public List<Map<String, Object>> rules(AuthenticatedUserPrincipal principal) { requireAdmin(principal); return ruleRepository.findAllByOrderByIdAsc().stream().map(this::toRule).toList(); }
    public Map<String, Object> createRule(AuthenticatedUserPrincipal principal, CompatibilityRuleRequest request) {
        requireAdmin(principal);
        PcCompatibilityRule rule = new PcCompatibilityRule();
        rule.setName(request.name());
        rule.setSourcePartType(request.sourcePartType());
        rule.setTargetPartType(request.targetPartType());
        rule.setRuleType(request.ruleType());
        rule.setRuleValueJson(writeJson(request.ruleValue()));
        return toRule(ruleRepository.save(rule));
    }
    public Map<String, Object> updateRule(AuthenticatedUserPrincipal principal, Long id, CompatibilityRuleRequest request) {
        requireAdmin(principal);
        PcCompatibilityRule rule = ruleRepository.findById(id).orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "호환성 규칙을 찾을 수 없습니다."));
        rule.setName(request.name());
        rule.setSourcePartType(request.sourcePartType());
        rule.setTargetPartType(request.targetPartType());
        rule.setRuleType(request.ruleType());
        rule.setRuleValueJson(writeJson(request.ruleValue()));
        return toRule(rule);
    }
    public Map<String, Object> deleteRule(AuthenticatedUserPrincipal principal, Long id) { requireAdmin(principal); ruleRepository.delete(ruleRepository.findById(id)
            .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "호환성 규칙을 찾을 수 없습니다."))); return Map.of("message", "호환성 규칙이 삭제되었습니다."); }

    private PcBuild requireBuild(Long id) { return pcBuildRepository.findById(id).orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "견적을 찾을 수 없습니다.")); }
    private PcBuild requireOwnedBuild(AuthenticatedUserPrincipal principal, Long id) {
        User user = requireCurrentUser(principal);
        return pcBuildRepository.findByIdAndUserId(id, user.getId()).orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "견적을 찾을 수 없습니다."));
    }
    private User requireCurrentUser(AuthenticatedUserPrincipal principal) {
        if (principal == null) { throw new BusinessException(ErrorCode.UNAUTHORIZED, "인증이 필요합니다."); }
        return userRepository.findById(principal.userId()).orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "사용자를 찾을 수 없습니다."));
    }
    private void requireAdmin(AuthenticatedUserPrincipal principal) {
        if (principal == null) { throw new BusinessException(ErrorCode.UNAUTHORIZED, "인증이 필요합니다."); }
        if (!"ADMIN".equalsIgnoreCase(principal.role())) { throw new BusinessException(ErrorCode.FORBIDDEN, "관리자 권한이 필요합니다."); }
    }
    private Map<String, Object> page(List<Map<String, Object>> items) { return Map.of("items", items, "pagination", Map.of("page", 1, "limit", items.isEmpty() ? 20 : items.size(), "total", items.size(), "totalPages", items.isEmpty() ? 0 : 1)); }
    private Map<String, Object> toSummary(PcBuild build) {
        Map<String, Object> item = new LinkedHashMap<>();
        item.put("id", build.getId());
        item.put("name", build.getName());
        item.put("description", build.getDescription() == null ? "" : build.getDescription());
        item.put("shareCode", build.getShareCode() == null ? "" : build.getShareCode());
        item.put("viewCount", build.getViewCount());
        item.put("createdAt", build.getCreatedAt() == null ? null : build.getCreatedAt().toString());
        item.put("updatedAt", build.getUpdatedAt() == null ? null : build.getUpdatedAt().toString());
        return item;
    }
    private Map<String, Object> toDetail(PcBuild build) {
        List<Map<String, Object>> parts = pcBuildPartRepository.findByPcBuildIdOrderByIdAsc(build.getId()).stream().map(part -> {
            Map<String, Object> product = new LinkedHashMap<>();
            product.put("id", part.getProduct().getId());
            product.put("name", part.getProduct().getName());
            product.put("slug", part.getProduct().getSlug());
            Map<String, Object> item = new LinkedHashMap<>();
            item.put("id", part.getId());
            item.put("partType", part.getPartType());
            item.put("quantity", part.getQuantity());
            item.put("product", product);
            return item;
        }).toList();
        Map<String, Object> response = new LinkedHashMap<>(toSummary(build));
        response.put("userId", build.getUser().getId());
        response.put("parts", parts);
        return response;
    }
    private Map<String, Object> toRule(PcCompatibilityRule rule) {
        Map<String, Object> item = new LinkedHashMap<>();
        item.put("id", rule.getId());
        item.put("name", rule.getName());
        item.put("sourcePartType", rule.getSourcePartType());
        item.put("targetPartType", rule.getTargetPartType());
        item.put("ruleType", rule.getRuleType());
        item.put("ruleValue", readJson(rule.getRuleValueJson()));
        return item;
    }
    private String writeJson(Map<String, Object> map) { try { return map == null ? null : objectMapper.writeValueAsString(map); } catch (JsonProcessingException e) { throw new BusinessException(ErrorCode.INTERNAL_ERROR, "호환성 규칙 저장에 실패했습니다."); } }
    private Map<String, Object> readJson(String value) { try { return value == null || value.isBlank() ? Map.of() : objectMapper.readValue(value, new TypeReference<Map<String, Object>>() {}); } catch (JsonProcessingException e) { throw new BusinessException(ErrorCode.INTERNAL_ERROR, "호환성 규칙을 읽을 수 없습니다."); } }
}

File: BackEnd/PHP/php-laravel-composer-eloquent-postgresql/app/Modules/PcBuilder/Services/PcBuilderService.php
<?php

namespace App\Modules\PcBuilder\Services;

use App\Common\Exceptions\BusinessException;
use App\Models\PcBuild;
use App\Models\PcBuildPart;
use App\Models\PcCompatibilityRule;
use App\Models\Product;
use App\Models\User;
use Illuminate\Support\Str;
use Symfony\Component\HttpFoundation\Response;

class PcBuilderService
{
    public function listMine(User $user, int $page, int $limit): array
    {
        $result = PcBuild::query()->where('user_id', $user->id)->orderByDesc('id')->paginate($limit, ['*'], 'page', $page);

        return [
            'items' => $result->getCollection()->map(fn (PcBuild $build): array => $this->serializeBuildSummary($build))->all(),
            'pagination' => $this->pagination($result),
        ];
    }

    public function create(User $user, array $payload): array
    {
        $build = PcBuild::query()->create([
            'user_id' => $user->id,
            'name' => $payload['name'],
            'description' => $payload['description'] ?? null,
        ]);

        return $this->serializeBuildDetail($build->fresh());
    }

    public function show(int $id): array
    {
        $build = PcBuild::query()->with('parts.product')->find($id);
        if (! $build) {
            throw new BusinessException('견적을 찾을 수 없습니다.', 'PC_BUILD_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $build->increment('view_count');

        return $this->serializeBuildDetail($build->fresh('parts.product'));
    }

    public function update(User $user, int $id, array $payload): array
    {
        $build = $this->findOwnedBuild($user, $id);
        $build->forceFill([
            'name' => $payload['name'] ?? $build->name,
            'description' => $payload['description'] ?? $build->description,
        ])->save();

        return $this->serializeBuildDetail($build->fresh('parts.product'));
    }

    public function delete(User $user, int $id): array
    {
        $build = $this->findOwnedBuild($user, $id);
        $build->delete();

        return ['message' => '견적이 삭제되었습니다.'];
    }

    public function addPart(User $user, int $buildId, array $payload): array
    {
        $build = $this->findOwnedBuild($user, $buildId);
        $product = Product::query()->find($payload['productId']);
        if (! $product) {
            throw new BusinessException('상품을 찾을 수 없습니다.', 'PRODUCT_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        PcBuildPart::query()->create([
            'pc_build_id' => $build->id,
            'part_type' => $payload['partType'],
            'product_id' => $product->id,
            'quantity' => (int) ($payload['quantity'] ?? 1),
        ]);

        return $this->serializeBuildDetail($build->fresh('parts.product'));
    }

    public function removePart(User $user, int $buildId, int $partId): array
    {
        $build = $this->findOwnedBuild($user, $buildId);
        $part = PcBuildPart::query()->where('pc_build_id', $build->id)->find($partId);
        if (! $part) {
            throw new BusinessException('부품을 찾을 수 없습니다.', 'PC_BUILD_PART_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $part->delete();

        return $this->serializeBuildDetail($build->fresh('parts.product'));
    }

    public function compatibility(int $buildId): array
    {
        $build = PcBuild::query()->with('parts.product')->find($buildId);
        if (! $build) {
            throw new BusinessException('견적을 찾을 수 없습니다.', 'PC_BUILD_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $issues = [];
        $parts = $build->parts;
        foreach (PcCompatibilityRule::query()->get() as $rule) {
            $source = $parts->firstWhere('part_type', $rule->source_part_type);
            $target = $parts->firstWhere('part_type', $rule->target_part_type);
            if (! $source || ! $target) {
                continue;
            }
            if ($rule->rule_type === 'CATEGORY_MISMATCH' && $source->product?->category_id === $target->product?->category_id) {
                $issues[] = [
                    'ruleId' => $rule->id,
                    'message' => $rule->name,
                    'severity' => 'WARNING',
                ];
            }
        }

        return ['compatible' => $issues === [], 'issues' => $issues];
    }

    public function share(User $user, int $buildId): array
    {
        $build = $this->findOwnedBuild($user, $buildId);
        if (! $build->share_code) {
            $build->forceFill(['share_code' => Str::lower(Str::random(12))])->save();
        }

        return [
            'shareUrl' => '/api/v1/pc-builds/shared/'.$build->share_code,
            'shareCode' => $build->share_code,
        ];
    }

    public function showShared(string $shareCode): array
    {
        $build = PcBuild::query()->with('parts.product')->where('share_code', $shareCode)->first();
        if (! $build) {
            throw new BusinessException('공유 견적을 찾을 수 없습니다.', 'PC_BUILD_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        return $this->serializeBuildDetail($build);
    }

    public function popular(int $page, int $limit): array
    {
        $result = PcBuild::query()->orderByDesc('view_count')->orderByDesc('id')->paginate($limit, ['*'], 'page', $page);

        return [
            'items' => $result->getCollection()->map(fn (PcBuild $build): array => $this->serializeBuildSummary($build))->all(),
            'pagination' => $this->pagination($result),
        ];
    }

    public function listRules(User $user): array
    {
        $this->assertAdmin($user);

        return PcCompatibilityRule::query()->orderBy('id')->get()->map(fn (PcCompatibilityRule $rule): array => $this->serializeRule($rule))->all();
    }

    public function createRule(User $user, array $payload): array
    {
        $this->assertAdmin($user);

        $rule = PcCompatibilityRule::query()->create([
            'name' => $payload['name'],
            'source_part_type' => $payload['sourcePartType'],
            'target_part_type' => $payload['targetPartType'],
            'rule_type' => $payload['ruleType'],
            'rule_value' => $payload['ruleValue'] ?? null,
        ]);

        return $this->serializeRule($rule);
    }

    public function updateRule(User $user, int $id, array $payload): array
    {
        $this->assertAdmin($user);
        $rule = PcCompatibilityRule::query()->find($id);
        if (! $rule) {
            throw new BusinessException('호환성 규칙을 찾을 수 없습니다.', 'COMPATIBILITY_RULE_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $rule->forceFill([
            'name' => $payload['name'] ?? $rule->name,
            'source_part_type' => $payload['sourcePartType'] ?? $rule->source_part_type,
            'target_part_type' => $payload['targetPartType'] ?? $rule->target_part_type,
            'rule_type' => $payload['ruleType'] ?? $rule->rule_type,
            'rule_value' => array_key_exists('ruleValue', $payload) ? $payload['ruleValue'] : $rule->rule_value,
        ])->save();

        return $this->serializeRule($rule);
    }

    public function deleteRule(User $user, int $id): array
    {
        $this->assertAdmin($user);
        $rule = PcCompatibilityRule::query()->find($id);
        if (! $rule) {
            throw new BusinessException('호환성 규칙을 찾을 수 없습니다.', 'COMPATIBILITY_RULE_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        $rule->delete();

        return ['message' => '호환성 규칙이 삭제되었습니다.'];
    }

    private function findOwnedBuild(User $user, int $id): PcBuild
    {
        $build = PcBuild::query()->with('parts.product')->where('user_id', $user->id)->find($id);
        if (! $build) {
            throw new BusinessException('견적을 찾을 수 없습니다.', 'PC_BUILD_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }

        return $build;
    }

    private function serializeBuildSummary(PcBuild $build): array
    {
        return [
            'id' => $build->id,
            'name' => $build->name,
            'description' => $build->description,
            'shareCode' => $build->share_code,
            'viewCount' => $build->view_count,
            'createdAt' => optional($build->created_at)?->toISOString(),
            'updatedAt' => optional($build->updated_at)?->toISOString(),
        ];
    }

    private function serializeBuildDetail(PcBuild $build): array
    {
        return array_merge($this->serializeBuildSummary($build), [
            'userId' => $build->user_id,
            'parts' => $build->parts->map(fn (PcBuildPart $part): array => [
                'id' => $part->id,
                'partType' => $part->part_type,
                'quantity' => $part->quantity,
                'product' => $part->product ? [
                    'id' => $part->product->id,
                    'name' => $part->product->name,
                    'slug' => $part->product->slug,
                ] : null,
            ])->all(),
        ]);
    }

    private function serializeRule(PcCompatibilityRule $rule): array
    {
        return [
            'id' => $rule->id,
            'name' => $rule->name,
            'sourcePartType' => $rule->source_part_type,
            'targetPartType' => $rule->target_part_type,
            'ruleType' => $rule->rule_type,
            'ruleValue' => $rule->rule_value,
            'createdAt' => optional($rule->created_at)?->toISOString(),
            'updatedAt' => optional($rule->updated_at)?->toISOString(),
        ];
    }

    private function pagination($result): array
    {
        return [
            'page' => $result->currentPage(),
            'limit' => $result->perPage(),
            'total' => $result->total(),
            'totalPages' => $result->lastPage(),
        ];
    }

    private function assertAdmin(User $user): void
    {
        if ($user->role !== 'ADMIN') {
            throw new BusinessException('관리자 권한이 필요합니다.', 'FORBIDDEN', Response::HTTP_FORBIDDEN);
        }
    }
}
